webapp/src/main/java/com/box/l10n/mojito/service/smartling/ThirdPartyTextUnitMatchingService.java

```diff
@@ -0,0 +1,173 @@
+
+    private Boolean checkForPartialMatchThirdPartyTextUnit(ThirdPartyTextUnit thirdPartyTextUnit) {
+        ThirdPartyTextUnit existingThirdPartyTextUnitPartialMatch = thirdPartyTextUnitRepository.findByThirdPartyTextUnitId(
+                thirdPartyTextUnit.getThirdPartyTextUnitId());
+        if (existingThirdPartyTextUnitPartialMatch != null) {
+            thirdPartyTextUnitRepository.delete(existingThirdPartyTextUnitPartialMatch);
+            thirdPartyTextUnitRepository.flush();
+        }
+        return existingThirdPartyTextUnitPartialMatch != null;
+    }
+
+    private Boolean matchWithMojitoTextUnit(ThirdPartyTextUnit thirdPartyTextUnit, String name, String assetPath) {
+        TMTextUnit matchingMojitoTextUnit = tmTextUnitRepository.findFirstByNameAndAssetPath(name, assetPath);
+        if (matchingMojitoTextUnit == null) {
+            // If no matching mojito text unit, check to see if it is a plural
+            // smartling sends plurals without the plural suffix
+            matchingMojitoTextUnit = tmTextUnitRepository.findFirstByNameAndAssetPath(name + "_other", assetPath);
+        }
+        thirdPartyTextUnit.setTmTextUnit(matchingMojitoTextUnit);
+        return matchingMojitoTextUnit != null;
+    }
+
+
+    private void matchSmartlingThirdPartyTextUnits(List<StringData> stringDataList, String file) {
+        Instant start = Instant.now();
+        AtomicInteger fullMatch = new AtomicInteger(0);
+        AtomicInteger partialMatch = new AtomicInteger(0);
+        AtomicInteger newThirdPartyTextUnit = new AtomicInteger(0);
+        AtomicInteger newTPTWithoutMatchingMojitoTU = new AtomicInteger(0);
+        AtomicInteger missingVariant = new AtomicInteger(0);
+        String delimiter = "#@#";
+        // partition the list of StringData into groups of 500
+        List<List<StringData>> batchedStringData = Lists.partition(stringDataList, 500);
+
+        batchedStringData.forEach(
+                stringDataBatch -> {
+                    Set<ThirdPartyTextUnit> batchInsertSet = new HashSet<>();
+                    stringDataBatch.forEach(
+                            stringData ->
+                                    stringData.getItems().forEach(
+                                            stringdebug -> {
+                                                String hashcode = stringdebug.getHashcode();
+                                                String mappingKey = stringdebug.getStringVariant();
+
+                                                if (mappingKey != null) {
+                                                    String[] mappingKeySplit = mappingKey.split(delimiter);
+                                                    String assetPath = mappingKeySplit[0];
+                                                    String name = mappingKeySplit[1];
+
+                                                    // Check to see if we already have a match for this hashcode and mapping key, if so we can skip it
```

Comment:
I am worried about performances here as it is making 1 query per text unit. In other places we batch fetch the info to perform the mapping. I'd have to look at it in more detail, just a first feedback.

Reply:
will await your further review

Reply:
I have now refactored the service